### PR TITLE
feat: Mosaic grid

### PR DESCRIPTION
**Pull request by richburdon**
Screenshot 2023-10-01 at 12 44 50 PM
Screenshot 2023-10-01 at 12 44 31 PM

### Summary
🚚🎨📝

This pull request refactors the aurora-grid package, mainly the mosaic component and its related hooks, types, and utilities. It separates the mosaic logic from the drag and drop logic, using a custom `MosaicDndProvider` component and a `MosaicDndContext` type. It also improves the readability, clarity, and documentation of the code, fixes some minor issues and typos, and adds a new property for the overlay drop animation.

> _Sing, O Muse, of the mighty refactor of the mosaic code,_
> _How the skilled developers rearranged the modules and types,_
> _And how they added new features and fixed the bugs they found,_
> _Like Hephaestus crafting wondrous works with fire and tools._

### Walkthrough
*  Refactor the `dnd` module to separate the mosaic logic from the DND logic and avoid circular dependencies ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-39095c3829d292d26fd4b508fa2d6d1620ec3d6773dadf7b02af2adfd8c4c585R85), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-39095c3829d292d26fd4b508fa2d6d1620ec3d6773dadf7b02af2adfd8c4c585R93), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-146dec3e137598e72cc87fafc4314bd7e423731063b4d4e700ffc339f36d5d39L7-R7), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6be316b4c5d3cde94edcc79f422839b1e2c98383285e8ea5e9f0a35f05d61a7aL9-R11), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-1adda71916bd5c607070350ec08279a385e2ae645a20dc118027318bd13d5798), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-5de3de90354f856d5fd4099088ecb373561e7b6ed31d5b09a5448085d5976043), F7L6L6, [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-92f1a516d20bc9edb4e7951d934154cca884b45605e70fb0cccaa41f51a357f3L16-R17), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-270384476efeacf5627da8a3dd668b50f8fe30f29485b960f7d4ec3830ba68c1L11-L13), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-270384476efeacf5627da8a3dd668b50f8fe30f29485b960f7d4ec3830ba68c1L24-R22), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-270384476efeacf5627da8a3dd668b50f8fe30f29485b960f7d4ec3830ba68c1L98-R98), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-77ba32845308a79a63bdc758d68969b77692cb6f496590752754612bade42a34), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-8bda7c636f76c32c023588ff2c59d28f549443a8c899b338c9c5ed1ff67f3093L8-R8), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-80fdf4eae8481cc0a2042219bebeb98a7457a1085e7d58d814bda5c4999fea44L10-R13), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-8206259edcaf7c9591167be63ab3591fc53174afbade8c389c2f382e040f8e09L8-R25), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-215d3b349c7a5aa68805ea487382a7c4df7f908b9f3ddd3f5b11e37503937a0dR1-R51), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-56294259a856ee8cfd902b7c933857ddc5fe3d2632fcd703c560018ce052750bR4), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6946803befd05be24ba6f76d2ed0391c382c98f0cb00e506c885e07b2e80b15dR9-R10), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6946803befd05be24ba6f76d2ed0391c382c98f0cb00e506c885e07b2e80b15dL71-R76), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-359349d98ce2bc1922332fcd7c2ba9f196ae586f803e64fbe5a0af74bfa3c710R8), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-0b102ca38b6ccfe522901b200e9301ee397ff0fd21f38b5a2dfef00cf18498f2L5-R8), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L10-R13))
  * Rename the `DndContext.tsx` file to `MosaicDndContext.tsx` and the `DndProvider` component to `MosaicDndProvider` to make them more specific and consistent ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L21-R29), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L56-R66), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L125-R130))
  * Move the type definition of `OverlayDropAnimation` from the `types.ts` file to the `drop-animations.ts` file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-146dec3e137598e72cc87fafc4314bd7e423731063b4d4e700ffc339f36d5d39L7-R7))
  * Remove some unused exports from the `dnd` index file and only export the utilities, the drop animations, and the mosaic DND context ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6be316b4c5d3cde94edcc79f422839b1e2c98383285e8ea5e9f0a35f05d61a7aL9-R11))
  * Delete the `types.ts` and `util` files from the `dnd` module, as they are no longer needed or moved to other modules ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-1adda71916bd5c607070350ec08279a385e2ae645a20dc118027318bd13d5798), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-5de3de90354f856d5fd4099088ecb373561e7b6ed31d5b09a5448085d5976043))
  * Update the imports in the `Mosaic.stories.tsx` file to reflect the new locations of some functions and types ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-92f1a516d20bc9edb4e7951d934154cca884b45605e70fb0cccaa41f51a357f3L16-R17))
  * Update the imports in the `Mosaic.tsx` file and replace the `DndProvider` component with the `MosaicDndProvider` component ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-270384476efeacf5627da8a3dd668b50f8fe30f29485b960f7d4ec3830ba68c1L11-L13), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-270384476efeacf5627da8a3dd668b50f8fe30f29485b960f7d4ec3830ba68c1L24-R22), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-270384476efeacf5627da8a3dd668b50f8fe30f29485b960f7d4ec3830ba68c1L98-R98))
  * Rename the `hooks` folder from the `dnd` module to the `mosaic` module and update the imports in the hook files ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-77ba32845308a79a63bdc758d68969b77692cb6f496590752754612bade42a34), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-8bda7c636f76c32c023588ff2c59d28f549443a8c899b338c9c5ed1ff67f3093L8-R8), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-80fdf4eae8481cc0a2042219bebeb98a7457a1085e7d58d814bda5c4999fea44L10-R13), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-8206259edcaf7c9591167be63ab3591fc53174afbade8c389c2f382e040f8e09L8-R25))
  * Move the `useHandleRearrange.ts` file from the `dnd` module to the `mosaic` module and add a new file `useHandleRearrange.ts` to the `mosaic` module ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-53ab4a2be52757814fe46eb5da2ad75d85b1cd2675fcbbdf5baf93e88770ee8f), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-215d3b349c7a5aa68805ea487382a7c4df7f908b9f3ddd3f5b11e37503937a0dR1-R51))
  * Add an empty line to the `Tile.tsx` file and add a comment to describe the `Tile` component ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-56294259a856ee8cfd902b7c933857ddc5fe3d2632fcd703c560018ce052750bR4), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-56294259a856ee8cfd902b7c933857ddc5fe3d2632fcd703c560018ce052750bL11-R17))
  * Add an import of the `Handler` type from the `dnd` module to the `types.ts` file and remove the `Handler` type definition from the `types.ts` file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6946803befd05be24ba6f76d2ed0391c382c98f0cb00e506c885e07b2e80b15dR9-R10), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6946803befd05be24ba6f76d2ed0391c382c98f0cb00e506c885e07b2e80b15dL71-R76))
  * Add an empty line to the `dnd-id.ts` file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-359349d98ce2bc1922332fcd7c2ba9f196ae586f803e64fbe5a0af74bfa3c710R8))
  * Add some new exports from the `util` module, which contains some helper functions that were moved from the `dnd` module or added to the `util` module ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-0b102ca38b6ccfe522901b200e9301ee397ff0fd21f38b5a2dfef00cf18498f2L5-R8))
  * Move the `manage-preview.ts` file from the `dnd` module to the `util` module and update the imports in the file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L10-R13))
* Improve the code readability and consistency in the `MosaicDndContext.tsx` and `useHandleMigrate.ts` files by removing the `deps` variables and passing the dependencies directly to the `useCallback` hooks ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L113-R122), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-8206259edcaf7c9591167be63ab3591fc53174afbade8c389c2f382e040f8e09L94-R107))
* Reorder the properties of the `DndContextValue` type alphabetically and add the `inhibitMigrationDestinationId` and `migrationDestinationId` properties to the type definition and the default context value in the `MosaicDndContext.tsx` file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L36-R43), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52R49-R51))
* Add periods to the end of the comments in the `MosaicDndContext.tsx` and `manage-preview.ts` files ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L67-R72), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L40-R47), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L55-R55), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L70-R70), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L89-R91), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-d583c300413f0c1cbf8ddb740c444d979acc2dc731acb73e7e364d4a8bc23340L99-R99))
* Add a TODO comment to rename the `Handler` type to `MosaicHandler` in the `MosaicDndContext.tsx` file and to make the `Delegator` type generic in the `types.ts` file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-2f4522f87d1f5a86b9d5856f1fc132a4bc0f06f13a699ffab9d5a7a70b516a52L21-R29), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-6946803befd05be24ba6f76d2ed0391c382c98f0cb00e506c885e07b2e80b15dR48))
* Change the TODO comment in the `Tile.tsx` file to use 'item' instead of 'listitem' as a possible variant name ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-56294259a856ee8cfd902b7c933857ddc5fe3d2632fcd703c560018ce052750bL11-R17))
* Remove some empty lines in the `Mosaic.stories.tsx` file ([link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-92f1a516d20bc9edb4e7951d934154cca884b45605e70fb0cccaa41f51a357f3R27), [link](https://github.com/dxos/dxos/pull/4342/files?diff=unified&w=0#diff-92f1a516d20bc9edb4e7951d934154cca884b45605e70fb0cccaa41f51a357f3L151))

### CODE REVIEW

File: packages/ui/aurora-grid/src/dnd/MosaicContext.tsx [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+/**
+ * Root provider.
+ */
+export const MosaicContextProvider: FC<MosaicContextProviderProps> = ({
+  Component = DefaultComponent,
+  debug,
+  children,
+}) => {
+  const [delegators] = useState(
```
Comment (dmaretskyi) [inline packages/ui/aurora-grid/src/dnd/MosaicContext.tsx:38]:
```suggestion
  const [containers] = useState(
```

File: packages/ui/aurora-grid/src/dnd/types.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// - Tile: component that is rendered within the layout and is draggable.
+// - Item: datum represented by a Tile.
+
+export type MosaicDataItem = { id: string };
+
+export type MosaicDraggedItem = {
+  container: string;
+  item: MosaicDataItem;
+  position?: any; // Index or layout-specific positional information (stored separately from the item).
+};
```
Comment (dmaretskyi) [inline packages/ui/aurora-grid/src/dnd/types.ts:18]: You should be able to specify types for callbacks attached to specific containers

```suggestion
export type MosaicDraggedItem<TPosition = unknown> = {
  container: string;
  item: MosaicDataItem;
  position?: TPosition; // Index or layout-specific positional information (stored separately from the item).
};
```

File: packages/ui/aurora-grid/src/dnd/types.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Tile container.
+ */
+export type MosaicContainerProps<TData extends MosaicDataItem> = {
+  id: string;
+  Component?: MosaicTileComponent<TData>;
+
+  // TODO(burdon): Rename drag events.
+  onMoveItem?: (event: MosaicMoveEvent) => void;
```
Comment (dmaretskyi) [inline packages/ui/aurora-grid/src/dnd/types.ts:56]: you can try redefining the event type to have specific data and position types

example: https://github.com/dxos/dxos/blob/2f18e291944691b1af166a5af3e5712d417c5f4b/packages/core/halo/credentials/src/credentials/assertions.ts#L26

File: packages/ui/aurora-grid/src/components/Grid/Grid.tsx [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+              />
+            );
+          })}
+        </div>
+      </div>
+    </div>
+  );
+};
+
+const Tile: FC<{
```
Comment (dmaretskyi) [inline packages/ui/aurora-grid/src/components/Grid/Grid.tsx:193]: Looks very similar to Stack tile..

File: packages/ui/aurora-grid/src/dnd/MosaicContext.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                id: overItem?.item?.id,
+                container: overItem?.container,
+              },
+            }}
+          />
+        )}
+
+        {/* Active dragging element. */}
+        {createPortal(
+          <DragOverlay>{activeItem && <OverlayComponent data={activeItem.item} isActive={true} />}</DragOverlay>,
```
Comment (dmaretskyi) [inline packages/ui/aurora-grid/src/dnd/MosaicContext.tsx:119]: morph based on `over` container?

File: packages/ui/aurora-grid/src/dnd/types.ts (hunk 3 above)
Comment (richburdon) [reply packages/ui/aurora-grid/src/dnd/types.ts:55]: let's have a conversation about generics. i think the general issue is:
the framework "routes" providers to consumers; both sides understand their own types -- but the framework has an `any` map. types are useful for casting, but the framework can't guarantee the type it's giving you. so isn't it just a cast?

File: packages/ui/aurora-grid/src/dnd/MosaicContext.tsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  if (activeItem && activeItem.container === id && overItem?.container !== activeItem.container) {
+    return items.filter((item) => item.id !== activeItem.item.id);
+  }
+  return items;
+};
+
+/**
+ * Register a container?
+ */
+export const useMosaicContainer = (container: MosaicContainerProps<any>) => {
```
[the remark this answers is not in the document]
Comment (richburdon) [reply packages/ui/aurora-grid/src/dnd/MosaicContext.tsx:155]: help!

File: packages/ui/aurora-grid/src/dnd/MosaicContext.tsx (hunk 5 above)
Comment (richburdon) [reply packages/ui/aurora-grid/src/dnd/MosaicContext.tsx]: that's working now. had to battle some DOM issues.

**Pull request merged**